packages/components/src/mobile/bottom-sheet-v2/index.native.js

```diff
@@ -0,0 +1,138 @@
+/**
+ * External dependencies
+ */
+import BottomSheetExternal, {
+	BottomSheetBackdrop,
+	BottomSheetScrollView,
+	useBottomSheetDynamicSnapPoints,
+} from '@gorhom/bottom-sheet';
+import { Modal } from 'react-native';
+import { GestureHandlerRootView } from 'react-native-gesture-handler';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	forwardRef,
+	useCallback,
+	useState,
+	useImperativeHandle,
+	useRef,
+} from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import styles from './style';
+
+function BottomSheetWithRef(
+	{ children, index, onClose, snapPoints = [ '50%' ] } = {},
+	ref
+) {
+	const renderBackdrop = useCallback(
+		( props ) => (
+			<BottomSheetBackdrop
+				{ ...props }
+				disappearsOnIndex={ -1 }
+				appearsOnIndex={ 0 }
+			/>
+		),
+		[]
+	);
+
+	const {
+		animatedHandleHeight,
+		animatedSnapPoints,
+		animatedContentHeight,
+		handleContentLayout,
+	} = useBottomSheetDynamicSnapPoints( snapPoints );
+
+	return (
+		<BottomSheetExternal
+			backdropComponent={ renderBackdrop }
+			backgroundStyle={ styles[ 'bottom-sheet-v2__background' ] }
+			enablePanDownToClose={ true }
+			contentHeight={ animatedContentHeight }
+			handleHeight={ animatedHandleHeight }
+			handleIndicatorStyle={
+				styles[ 'bottom-sheet-v2__handle-indicator' ]
+			}
+			index={ index }
+			onClose={ onClose }
+			ref={ ref }
+			snapPoints={ animatedSnapPoints }
+		>
+			<BottomSheetScrollView onLayout={ handleContentLayout }>
+				{ children }
+			</BottomSheetScrollView>
+		</BottomSheetExternal>
+	);
+}
+
+const BottomSheet = forwardRef( BottomSheetWithRef );
+
+const BottomSheetModalWithRef = (
+	{ index = 0, onClose, ...bottomSheetProps },
+	ref
+) => {
+	const bottomSheetRef = useRef( null );
+
+	/**
+	 * `internalIndex` is used to allow displaying the modal on initial render,
+	 * which is required in some areas of the code base that do not easily support
+	 * an call to an imperative `present` method.
+	 */
+	const [ internalIndex, setInternalIndex ] = useState( index );
+	const [ visible, setVisible ] = useState( index >= 0 );
+
+	const handlePresent = useCallback( () => {
+		setVisible( true );
+		setInternalIndex( index >= 0 ? index : 0 );
+	}, [] );
+
+	const handleDismiss = useCallback( () => {
+		bottomSheetRef.current?.close();
+	}, [] );
+
+	/**
+	 * Utilize imperative handle to mimic the `@gorhom/bottom-sheet` API, which
+	 * would simplify migrating to `BottomSheetModal` in the future if the editor
+	 * header navigation is rendered by React Native, not the native host app.
+	 */
+	useImperativeHandle( ref, () => ( {
+		present: handlePresent,
+		dismiss: handleDismiss,
+	} ) );
+
+	const handleClose = useCallback( () => {
+		setVisible( false );
+		if ( onClose ) {
+			onClose();
+		}
+	}, [ onClose ] );
+
+	return (
+		<Modal
```

Comment:
`Modal` is used rather than `@gorhom/bottom-sheet`'s `BottomSheetModal` as the latter is incapable of rendering atop the WP host app header navigation. If the header navigation is ever managed by React Native, then we may be able to switch modal providers.